kasa/smartplug.py

```diff
@@ -89,3 +89,8 @@ def state_information(self) -> Dict[str, Any]:
         """Return switch-specific state information."""
         info = {"LED state": self.led, "On since": self.on_since}
         return info
+
+    @property
+    def has_led(self) -> bool:
+        """Return True if device has an led switch."""
+        return True
```

Comment:
I think we should avoid doing this and keep the API clean. What I think we could use the opportunity to clean up some antics (`features`, `state_information`, ..) and introduce a layer of abstraction here:
1. We define an interface to let device implementations define individual features (maybe this could be called a `Feature`, naming things is hard..)
2. This interface defines _metadata_ about the exposed information, for the `LED` feature, we would add `name` (to be used in the cli), `icon` (for homeassistant), `setter` (if this is a changeable setting and not just a sensor value), category of the feature (informational, debug, or so, something to map to `entity_category` in homeassistant), etc.
3. In the cli and in the  homeassistant integration, we don't hard code the sensors, switches, etc. but use these "descriptors" to construct the wanted output (or entities, in homeassistant).

This will make introducing new settings and sensors easy for downstreams, as there will be no need to keep up with hardcoding new features if and when they come up. This will be sort of necessary anyway if the this project grows to implement hub-style features, so I think it'd be a good idea to tackle this early on.

Shortly put: we define a set of rather strict, device-type specific interfaces (plugs, bulbs, dimmers, ..) which describe basic features (turning on/off, setting brightness, ..) but all extra will be exposed through a generic way.

How to name such is in the air, but I hope this gives an idea about what I'm thinking. Any thoughts? :-)

Reply:
This seems reasonable although the risk with trying to something generically is it can often become messy and too generic. Children and transitions are a couple of things that spring to mind as sufficiently distinct to make a metadata approach complex.  What if the features are a bit more concrete like SwitchFeature or TransitionFeature?

Regardless of that do we see something like this:

```
## library
interface Device:
  def Features[FeatureType]: -> Feature
...
interface Feature:
  name
  icon
...
abstract Switch(Feature)
  __init__(self, turn_on_method, turn_off_method, state_method)

  def turn_on()
    self.turn_on_method()
...
## consumer
device: IDevice = get_device_somehow()
if led_switch:= device.Features[SwitchType.LED]:
  create_some_entity(led_switch)
...
def turn_on():
  led_switch.on()

```

In the meantime to keep this PR from ballooning I'll remove the `has_led` and do a `has_attr()` in HA until we have the interfaces in.  Otherwise the TapoPlug doesn't work because it bugs out expecting `len_on` etc.

Reply:
`has_attr` approach sounds good enough to me. I don't think we should go that deep into the abstractions as your example, I was thinking something like creating something rather simple to store the attribute value & some metadata on it, which can also implement `__str__` to keep the console output as it was. I'll try to elaborate about that briefly below, the namings etc. are subject to change obviously.

So maybe something like this (using `TapoPlug` as an example):
```
class StateInfo:
    ... 
    def __str__(self):
        return str(self.value)

def state_information(self) -> Dict[str, StateInfo]:
    return {
        "auto_off": StateInfo(title="Turn device off automatically", value=.., setter="method_to_change_the_state", icon="mdi:sleep"),
        "auto_off_remaining_time": StateInfo(title="Auto off in", type=datetime.timedelta, value=.., icon="mdi:sleep")}),
        "led": StateInfo(title="LED", type=bool, setter="set_led", icon="mdi:led"),
        "on_since": StateInfo(title="On since", icon="mdi:hourglass", value=...),
    }
```

Now, this could also be extended to support the common functions (like on/off, color temp, ..) but it's just an idea that would make it rather simple to add new functionalities into homeassistant without writing some integration code and tests to homeassistant code when some new information is deemed interesting enough to be exposed.

https://github.com/home-assistant/core/pull/102183 is somewhat related, basically we could avoid much homeassistant boilerplate if we'd just define a data model with enough metadata to construct all of it automatically. RSSI is considered interesting information? Define it inside python-kasa, bump the version, and boom you have it :-)

Reply:
All makes sense.  Resolving this and linking to it in the discussion for future reference